File: Moon.CSharp/Component/Visual/SpriteDir.cs
using Fodot.CSharp;
using Fodot.Injection;
using Godot;
using Moon.CSharp;
using Moon.Interface;
using Moon.Library;

namespace Moon.Component;

[GlobalClass, ChildOf("CanvasItem")]
public partial class SpriteDir : Node, IFlipHInit
{
    /// <summary>
    /// The monitored moving node.
    /// </summary>
    [Export(PropertyHint.NodePathValidTypes, "CanvasItem")]
    public NodePath TrackingNode { get ;set; }
    
    [Export(PropertyHint.NodePathValidTypes, "Rotator")]
    public NodePath RotatorNode { get ;set; }
    
    [Export]
    public bool Flip { get ;set; }
    
    [Export]
    public bool Disabled { get ;set; }
    
    private CanvasItem Root;
    private CanvasItem Sprite;
    private Rotator Rotator;
    
    private Recorder2D Recorder;
    
    public override void _EnterTree()
    {
        base._EnterTree();
    #if DEBUG
        if (Engine.IsEditorHint()) return;
    #endif
    
        Sprite = GetParentOrNull<CanvasItem>();
    }
    
    public override void _Ready()
    {
        base._Ready();
    #if DEBUG
        if (Engine.IsEditorHint()) return;
    #endif
    
        Root = GetNodeOrNull<CanvasItem>(TrackingNode);
        Rotator = GetNodeOrNull<Rotator>(RotatorNode);
        
        if (IsInstanceValid(Root)) Recorder = Recorder2DModule.get(Root);
        this.AddPhysicsProcess(Process);
    }

    public void InitFlipH()
    {
        this.WhenReady(() => SetSpriteFlip(true));
    }
    
    private void Process()
    {
        if (IsInstanceValid(Root))
        {
            var s = Recorder.LastMotion.X;
            if (s != 0f) SetSpriteFlip(s < 0f);
        }
    }

    protected void SetSpriteFlip(bool value)
    {
        if (Disabled) return;
        
        var result = Flip ? !value : value;
        Sprite?.TrySetFlipH(result);
        Rotator?.Flip = result;
    }
}

File: Moon.Script/SpriteDir.fs
namespace Moon.Script

open Fodot
open Godot
open Moon.Component
open Moon.Interface
open Moon.Library
open Moon.Module

[<FScript(typeof<SpriteDir>)>]
type private SpriteDirScript(d : SpriteDir) =
    let spr = d |> Node.tryGetParent<CanvasItem>
    let mov = d |> Node.tryGetNode d.TrackingNode
    let rot = d |> Node.tryGetNode<Rotator> d.RotatorNode
    
    let recorder = mov |> Option.map Recorder2D.get
    
    let setFlip value =
        let result = if d.Flip then value |> not else value
        spr |> Option.iter (fun s -> s |> Flip.setH result)
        rot |> Option.iter (fun r -> r.Flip <- result)
        
    let update () =
        if d.Disabled then () else
        recorder
        |> Option.iter (fun r ->
            let s = r.LastMotion.X
            if s <> 0f then setFlip (s < 0f)
        )
        
    do d |> Engine.addPhysicsProcess update |> ignore
    
    interface IFlipHInit with
        member this.InitFlipH () =
            d |> Node.whenReady (fun _ -> setFlip true)
